sql/core/src/main/java/org/apache/spark/sql/sources/v2/WritableDataSourceV2.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2;
+
+import java.util.Map;
+
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.sources.v2.writer.DataSourceV2Writer;
+
+/**
+ * A mix in interface for `DataSourceV2`. Users can implement this interface to provide data writing
+ * ability with job level transaction. The writing procedure is:
+ *
+ *   1. create the writer, serialize and send it to all the partitions of the input data(RDD).
+ *   2. for each partition, write the data with the writer.
+ *   3. for each partition, if all data are written successfully, call writer.commit().
+ *   4. for each partition, if exception happens during the writing, call writer.abort().
+ *   5. if all partitions are written successfully, call WritableDataSourceV2.commit().
+ *   6. if some partitions are failed, call WritableDataSourceV2.abort().
+ *
+ * Note that, the implementations are responsible to correctly implement transaction by overwriting
+ * the `commit` and `abort` methods of the writer and this interface.
+ */
+public interface WritableDataSourceV2 extends DataSourceV2 {
+
+  /**
+   * The main entrance for write interface.
+   *
+   * @param mode the save move, can be append, overwrite, etc.
+   * @param options the options for this data source writer. These string options are
+   *                case-insensitive, which means, all the map keys are lower-cased, please be
+   *                careful when reading these options in the implementations.
+   * @return a writer that implements the actual write logic.
+   */
+  DataSourceV2Writer getWriter(SaveMode mode, Map<String, String> options);
+
+  // TODO: shall we force users to implement this?
```

Comment:
I'd force them to implement everything. When implementing a data source this is really trivial if it is no-op. Forcing them to implement it would make it less likely for somebody to forget.